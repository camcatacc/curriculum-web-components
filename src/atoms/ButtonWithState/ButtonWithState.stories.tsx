// Modules
import React, { useState } from "react";

// Storybook Modules
import { ComponentStory, ComponentMeta } from "@storybook/react";

// Elements
import ButtonWithState from "atoms/ButtonWithState/ButtonWithState";

// Template
const Template: ComponentStory<typeof ButtonWithState> = (args) => {
	const [loading, setLoading] = useState(false);

	function onClick() {
		setLoading(true);
		setLoadingFalseAfter2s();
	}

	const setLoadingFalseAfter2s = () => setTimeout(() => setLoading(false), 2000);

	return (
		<div>
			<ButtonWithState {...args} loading={loading} onClick={onClick} disabled={loading} />
		</div>
	);
};

// Variants
export const Standard = Template.bind({});
Standard.args = {
	variant: "contained",
	children: "Click me"
};

// Export
export default {
	title: "Atoms/ButtonWithState",
	component: ButtonWithState
} as ComponentMeta<typeof ButtonWithState>;
