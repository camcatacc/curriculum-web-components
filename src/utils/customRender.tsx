import { ThemeProvider } from "@material-ui/styles";
import { render, RenderOptions, RenderResult } from "@testing-library/react";
import { ReactElement } from "react";
import { getMaterialTheme } from "styles/materialUi";

export const customRender = (ui: ReactElement, renderOptions?: RenderOptions & { darkMode?: boolean }): RenderResult => {
	const Wrapper: React.FC = ({ children }) => {
		return <ThemeProvider theme={getMaterialTheme(renderOptions?.darkMode ? "dark" : "light")}>{children}</ThemeProvider>;
	};

	return render(ui, { wrapper: Wrapper, ...renderOptions });
};
