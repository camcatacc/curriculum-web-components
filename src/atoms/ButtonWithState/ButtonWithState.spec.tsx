// Modules
import React from "react";
import { fireEvent } from "@testing-library/react";
import { customRender } from "utils/customRender";
import "@testing-library/jest-dom/extend-expect";

// Elements
import ButtonWithState from "atoms/ButtonWithState/ButtonWithState";

// Definitions
import type { ButtonWithStateProps } from "atoms/ButtonWithState/ButtonWithState";

const defaultProps: ButtonWithStateProps = {
	children: "Loading button",
	onClick: jest.fn()
};
// Rendering
const renderComponent = (props = {}) => {
	const combinedProps = { ...defaultProps, ...props };
	return customRender(<ButtonWithState {...combinedProps} />);
};

describe("LoadingButton", () => {
	it("Snapshot", () => {
		const { container } = renderComponent();
		expect(container).toMatchSnapshot();
	});

	it("Calls onClick when clicked", () => {
		const { getByRole } = renderComponent();
		const button = getByRole("button");
		fireEvent.click(button);
		expect(defaultProps.onClick).toHaveBeenCalled();
	});

	it("Does not call onClick if disabled", () => {
		const { getByRole } = renderComponent({ disabled: true });
		const button = getByRole("button");
		fireEvent.click(button);
		expect(defaultProps.onClick).not.toHaveBeenCalled();
	});

	it("Shows loading if loading", () => {
		const { getByRole } = renderComponent({ loading: true });
		const loading = getByRole("progressbar");
		expect(loading).toBeInTheDocument();
	});
});
// Tests
