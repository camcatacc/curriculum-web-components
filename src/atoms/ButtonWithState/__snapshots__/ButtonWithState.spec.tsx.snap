// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingButton Snapshot 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-1"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSecondary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Loading button
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
